packages/config/src/layer2s/common/ProjectDiscovery.ts

```diff
@@ -0,0 +1,109 @@
+import { assert } from '@l2beat/common'
+import {
+  ContractParameters,
+  ContractValue,
+  ProjectParameters,
+} from '@l2beat/types'
+import { utils } from 'ethers'
+import fs from 'fs'
+import path from 'path'
+
+import { ProjectUpgradeability } from './../../common/ProjectContracts'
+
+type AllKeys<T> = T extends T ? keyof T : never
+
+type MergedUnion<T extends object> = {
+  [K in AllKeys<T>]: PickType<T, K>
+}
+
+type PickType<T, K extends AllKeys<T>> = T extends { [k in K]?: T[K] }
+  ? T[K]
+  : undefined
+
+export type Filesystem = typeof filesystem
+const filesystem = {
+  readFileSync: (path: string) => {
+    return fs.readFileSync(path, 'utf-8')
+  },
+}
+
+export class ProjectDiscovery {
+  private readonly discovery: ProjectParameters
+  constructor(
+    private readonly projectName: string,
+    private readonly fs: Filesystem = filesystem,
+  ) {
+    this.discovery = this.getDiscoveryJson(projectName)
+  }
+
+  private getDiscoveryJson(project: string): ProjectParameters {
+    const discoveryFile = this.fs.readFileSync(
+      path.resolve(`../backend/discovery/${project}/discovered.json`),
+    )
+
+    return JSON.parse(discoveryFile) as ProjectParameters
+  }
+
+  getContract(identifier: string): ContractParameters {
+    try {
+      identifier = utils.getAddress(identifier)
+    } catch {
+      return this.getContractByName(identifier)
+    }
+    return this.getContractByAddress(identifier)
+  }
+
+  getContractValue<T extends ContractValue>(
+    contractIdentifier: string,
+    key: string,
+  ): T {
+    const contract = this.getContract(contractIdentifier)
+    const result = contract.values?.[key] as T | undefined
```

Comment:
This is probably the least typeSafe part of this approach. Can we maybe check it somehow here?